subsys/net/lib/conn_mgr/conn_mgr.c

```diff
@@ -0,0 +1,197 @@
+	if (iface_states[index] & NET_STATE_CONNECTED) {
+		NET_DBG("Iface %p connected", iface);
+		net_mgmt_event_notify(NET_EVENT_L4_CONNECTED, iface);
+	} else {
+		NET_DBG("Iface %p disconnected", iface);
+		net_mgmt_event_notify(NET_EVENT_L4_DISCONNECTED, iface);
+	}
+}
+
+static void conn_mgr_act_on_changes(void)
+{
+	int idx;
+
+	for (idx = 0; idx < CONN_MGR_IFACE_MAX; idx++) {
+		enum net_conn_mgr_state state;
+
+		if (!(iface_states[idx] & NET_STATE_CHANGED)) {
+			continue;
+		}
+
+		state = NET_CONN_MGR_STATE_CONNECTED;
+
+		state &= conn_mgr_iface_status(idx);
+		if (state) {
+			enum net_conn_mgr_state ip_state =
+				NET_CONN_MGR_STATE_DISCONNECTED;
+
+			if (IS_ENABLED(CONFIG_NET_IPV6)) {
+				ip_state |= conn_mgr_ipv6_status(idx);
+			}
+
+			if (IS_ENABLED(CONFIG_NET_IPV4)) {
+				state |= conn_mgr_ipv4_status(idx);
+			}
+
+			state &= ip_state;
+		}
+
+		iface_states[idx] &= ~NET_STATE_CHANGED;
+
+		if (state == NET_CONN_MGR_STATE_CONNECTED &&
+		    !(iface_states[idx] & NET_STATE_CONNECTED)) {
+			iface_states[idx] |= NET_STATE_CONNECTED;
+
+			conn_mgr_notify_status(idx);
+		} else if (state != NET_CONN_MGR_STATE_CONNECTED &&
+			   (iface_states[idx] & NET_STATE_CONNECTED)) {
+			iface_states[idx] &= ~NET_STATE_CONNECTED;
+
+			conn_mgr_notify_status(idx);
+		}
+	}
+}
+
+static void conn_mgr_initial_state(struct net_if *iface)
+{
+	int idx = net_if_get_by_iface(iface) - 1;
+
+	if (net_if_is_up(iface)) {
+		NET_DBG("Iface %p UP", iface);
+		iface_states[idx] = NET_STATE_IFACE_UP;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV6)) {
+		if (net_if_ipv6_get_global_addr(NET_ADDR_PREFERRED,
+						&iface)) {
+			NET_DBG("IPv6 addr set");
+			iface_states[idx] |= NET_STATE_IPV6_ADDR_SET |
+				NET_STATE_IPV6_DAD_OK;
+		} else if (net_if_ipv6_get_global_addr(NET_ADDR_TENTATIVE,
+						       &iface)) {
+			iface_states[idx] |= NET_STATE_IPV6_ADDR_SET;
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV4)) {
+		if (net_if_ipv4_get_global_addr(iface,
+						NET_ADDR_PREFERRED)) {
+			NET_DBG("IPv4 addr set");
+			iface_states[idx] |= NET_STATE_IPV4_ADDR_SET;
+		}
+
+	}
+
+	iface_states[idx] |= NET_STATE_CHANGED;
+}
+
+static void conn_mgr_init_cb(struct net_if *iface, void *user_data)
+{
+	ARG_UNUSED(user_data);
+
+	conn_mgr_initial_state(iface);
+}
+
+static void conn_mgr(void)
+{
+	conn_mgr_init_events_handler();
+
+	net_if_foreach(conn_mgr_init_cb, NULL);
+
+	NET_DBG("Connection Manager started");
+
+	while (true) {
+		k_sem_take(&conn_mgr_lock, K_FOREVER);
+
+		conn_mgr_act_on_changes();
+	}
+}
+
+K_THREAD_DEFINE(conn_mgr_thread, CONFIG_NET_CONNECTION_MANAGER_STACK_SIZE,
+		(k_thread_entry_t)conn_mgr, NULL, NULL, NULL,
+		K_PRIO_COOP(2), 0, 0);
+
+void net_conn_mgr_resend_status(void)
+{
+	int idx;
+
+	for (idx = 0; idx < CONN_MGR_IFACE_MAX; idx++) {
+		conn_mgr_notify_status(idx);
+	}
+}
+
+static int conn_mgr_init(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	k_thread_start(conn_mgr_thread);
+
+	return 0;
+}
+
+SYS_INIT(conn_mgr_init, APPLICATION, 0);
```

Comment:
This should be settable via Kconfig file, we have it like that in all the other places in networking stack.

Reply:
ok